Mixed ER Pools

kbranch opened the pull request:
This allows simple entrances and connectors to be shuffled in a single pool.

I also added separate flags for watery entrances (Manbo, damp cave) and the old Insanity entrances (Mamu and raft house). There's plenty of room for subjectivity there, I'm happy to rework the settings. There's a lot of discussion in #91, but not really firm conclusions.

I think the new entrance swapping scheme might make [this loop](https://github.com/daid/LADXR/blob/master/randomizer.py#L42) unnecessary. I ran a bunch of tests without ever hitting a second iteration of the loop, but obviously I might've just not tested the right settings.

This branch is live here: https://kbranch.dev/LADXR/www/

kbranch commented:
Oh, and I did make a few attempts at being a little smarter about which entrances were switched, but every attempt ended up taking more iterations on average than the naive rnd.choice approach. There may still be some low hanging fruit for improvements there that I missed, though.

daid commented on worldSetup.py:
```diff
@@ -46,47 +47,78 @@ def getEntrancePool(self, settings, connectorsOnly=False):
+                self.entrance_mapping[entrance] = unmappedEntrances.pop(rnd.randrange(len(unmappedEntrances)))
+
+        # Make sure all entrances in the pool are accessible
+        for j in range(1000):
```
If we reach the end of a 1000 and not found a solution, we better throw an exception here. Instead of later during item placement.

daid commented on settings.py:
```diff
@@ -134,14 +134,18 @@ def __init__(self, multiworld_count=None):
+[Split] Connector caves are also randomized, in a separate pool from single entrance caves
+[Mixed] Connector caves are also randomized, in the same pool as single entrance caves
+If random start location and/or dungeon shuffle is enabled, then these will be shuffled with all the entrances."""),
+            Setting('shufflejunk', 'Gameplay', 'j', 'Shuffle itemless entrances', default=False,
```
With a bunch of extra ER related settings, it might be wise that we move all ER settings to it's own category instead of in 'Gameplay'?

daid replied:
With all these new settings, are there settings combinations that are not allowed?

kbranch commented:
> If we reach the end of a 1000 and not found a solution, we better throw an exception here. Instead of later during item placement.

Yep, good point, better to complain when we know what the problem is.

> With a bunch of extra ER related settings, it might be wise that we move all ER settings to it's own category instead of in 'Gameplay'?

Oh, yep, I overlooked these when I split the ones on the site out.

> With all these new settings, are there settings combinations that are not allowed?

The only thing I can think of is if any of the three new boolean settings are set and ER is set to default - in that case the new ones just do nothing. I just updated the descriptions to reflect that.

Otherwise, I think everything will operate pretty independently. The three new ones only modify which simple entrances are in the pool.

kbranch commented:
Turns out my hack to keep D7 from being a connector was half assed, I reworked it.

muffinjets commented:
Not to push anything before it's ready, but what's stopping this from being merged?

daid commented:
> Not to push anything before it's ready, but what's stopping this from being merged?

Me being lazy and distracted :)
